src/logging/httpLogger.ts

```diff
@@ -0,0 +1,119 @@
+import type {
+	InternalAxiosRequestConfig,
+	AxiosResponse,
+	AxiosError,
+} from "axios";
+import { isAxiosError } from "axios";
+import { getErrorMessage } from "coder/site/src/api/errors";
+import { getErrorDetail } from "../error";
+import {
+	formatHeaders,
+	formatBody,
+	formatUri,
+	formatContentLength,
+	formatMethod,
+} from "./formatters";
+import type { Logger } from "./logger";
+import {
+	HttpClientLogLevel,
+	RequestMeta,
+	RequestConfigWithMeta,
+} from "./types";
+import { shortId, formatTime, createRequestId } from "./utils";
+
+export function createRequestMeta(): RequestMeta {
+	return {
+		requestId: createRequestId(),
+		startedAt: Date.now(),
+	};
+}
+
+export function logRequest(
+	logger: Logger,
+	requestId: string,
+	config: InternalAxiosRequestConfig,
+	logLevel: HttpClientLogLevel,
+): void {
+	if (logLevel === HttpClientLogLevel.NONE) {
+		return;
+	}
+
+	const method = formatMethod(config.method);
+	const url = formatUri(config);
+	const len = formatContentLength(config.headers);
+
+	let msg = `→ ${shortId(requestId)} ${method} ${url} ${len}`;
+	if (logLevel >= HttpClientLogLevel.HEADERS) {
+		msg += `\n${formatHeaders(config.headers)}`;
+	}
+
+	if (logLevel >= HttpClientLogLevel.BODY) {
+		msg += `\n${formatBody(config.data)}`;
+	}
+
+	logger.trace(msg);
+}
+
+export function logResponse(
+	logger: Logger,
+	meta: RequestMeta,
+	response: AxiosResponse,
+	logLevel: HttpClientLogLevel,
+): void {
+	if (logLevel === HttpClientLogLevel.NONE) {
+		return;
+	}
+
+	const method = formatMethod(response.config.method);
+	const url = formatUri(response.config);
+	const time = formatTime(Date.now() - meta.startedAt);
+	const len = formatContentLength(response.headers);
+
+	let msg = `← ${shortId(meta.requestId)} ${response.status} ${method} ${url} ${len} ${time}`;
+
+	if (logLevel >= HttpClientLogLevel.HEADERS) {
+		msg += `\n${formatHeaders(response.headers)}`;
+	}
+
+	if (logLevel >= HttpClientLogLevel.BODY) {
+		msg += `\n${formatBody(response.data)}`;
+	}
+
+	logger.trace(msg);
+}
+
+export function logRequestError(
+	logger: Logger,
+	error: AxiosError | unknown,
+): void {
+	if (isAxiosError(error)) {
+		const config = error.config as RequestConfigWithMeta | undefined;
+		const meta = config?.metadata;
+		const method = formatMethod(config?.method);
+		const url = formatUri(config);
+		const requestId = meta?.requestId || "unknown";
+		const time = meta ? formatTime(Date.now() - meta.startedAt) : "?ms";
+
+		const msg = getErrorMessage(error, "No error message");
+		const detail = getErrorDetail(error) ?? "";
+
+		if (error.response) {
+			const responseData =
+				error.response.statusText || String(error.response.data).slice(0, 100);
```

Comment:
After experimenting a bit, I wonder if it is too much to show both the error detail and the status text?  Maybe we only show status text if the msg and detail are blank.

For example I can get this somewhat redundant line:

> 2025-09-15 13:37:29.711 [error] ← 9253a837 401 GET /api/v2/users/me 328ms - You are signed out or your session has expired. Please sign in again to continue. - API key expired at "2025-09-09 17:03:07.82664 +0000 UTC". - Unauthorized Request failed with status code 401